### ISSUE TITLE
Is there a way of adding a REST API call that causes falcon to cleanly shut down?

### ISSUE THREAD

**Issue by dstromberg** [falconry/falcon#1559]
Hi.

I'm working on getting coverage information out of a REST API written using Python+Falcon.

To do this, I'm using coverage.py.  It appears that coverage.py doesn't like to do its analysis on processes that never terminate; I think it wants to write the .coverage file on process exit.   

Also, it seems like falcon doesn't want to shut down.  To deal with that, I'd like to add an API call that cleanly shuts down the API I'm testing.  So I would fire up the API, run the automated tests, issue the terminate call to the API, and collect the coverage results.

Does falcon have some sort of "do a clean shutdown" callable somewhere?  I grep'd the sources a little, and didn't find anything.

If not, could one be added?

Thanks for the cool software.  :)

**Comment by dstromberg** [comment]
We're starting the API with:
```
if __name__ == '__main__':
    httpd = simple_server.make_server(api_config['API_BIND_ADDRESS'], int(api_config['API_BIND_PORT']), app)
    httpd.serve_forever()
```

**Comment by dstromberg** [comment]
It looks kinda like serve_forever() is coming from /usr/lib/python2.7/SocketServer.py (on CPython 2.7, on a Linux Mint 19.1 system), which has a __shutdown_request boolean that is set True by a shutdown() method.  I may experiment with that.

**Comment by kgriffs** [comment]
Hi @dstromberg, since Falcon simply exposes a WSGI callable that the server calls for each request, lifecycle management of the server process is something the framework can not influence. In the case of wsgiref, you can see how we do that for one of Falcon's own tests here:

https://github.com/falconry/falcon/blob/master/tests/test_wsgi.py

Alternatively, you can use Falcon's testing framework to simulate requests, which would still allow you to verify coverage. The tutorial in the docs includes a session on testing that demonstrates how this works:

https://falcon.readthedocs.io/en/stable/user/tutorial.html#testing-your-application

**Comment by dstromberg** [comment]
I managed to get a shutdown request working by duplicating a little bit of:
`/usr/lib/python2.7/SocketServer.py`'s `SockerServer.BaseServer.serve_forever` to add my own shutdown flag.

The chief disadvantage was that I needed to call `self._handle_request_noblock()`, which of course is a "hands mostly off" method.

So I'm only doing this when we want to collect test coverage data, not in production.

I needed to duplicate it, IIRC, because shutting down the API from within an API call would hang, but by splitting the duplicated code out into a standalone function, there's no hang.

Although I've referenced CPython 2.7 above, I'm mostly using it on CPython 3.x.
